pyro/infer/autoguide/guides.py

```diff
@@ -583,6 +583,133 @@ def _loc_scale(self, *args, **kwargs):
+        See :ref:`autoguide-initialization` section for available functions.
+    """
+
+    def __init__(self, model, prefix="auto", init_loc_fn=init_to_median, rank=1):
+        if not isinstance(rank, int) or not rank > 0:
+            raise ValueError("Expected rank > 0 but got {}".format(rank))
+        self.rank = rank
+        self._init_scale = 0.
+        self._init_cov = 0.
+        super(AutoProgressive, self).__init__(model, prefix=prefix, init_loc_fn=init_loc_fn)
+
+    def init_scale(self, scale=0.1):
+        """
+        Initialize diagonal uncertainty scale.
+        This is typically called after a point estimate has been learned.
+
+        :param float scale: An initial uncertainty scale. Must be postive.
+            Values less than 1 are recommended.
+        """
+        if not isinstance(scale, float) or not scale > 0.:
+            raise ValueError("Expected positive init_scale but got {}".format(scale))
+        if self._init_scale > 0.:
+            raise ValueError("Cannot initialize scale a second time.")
+        self._init_scale = scale
+
+    def init_cov(self, scale=0.1):
+        """
+        Initialize low rank covariance factor to scaled random noise.
+        This is typically called after both point estimates and noise scales
+        have been learned; the initial low rank ``cov_factor`` scale combines
+        the ``scale`` arg with the current learned diagonal noise scales.
+
+        :param float scale: A noise scale. Must be positive.
+            Values less than 1 are recommended.
+        """
+        if not isinstance(scale, float) or not scale > 0.:
+            raise ValueError("Expected positive init_cov but got {}".format(scale))
+        if self._init_cov > 0.:
+            raise ValueError("Cannot initialize cov_factor a second time.")
+        if self._init_scale == 0.:
+            raise ValueError("Cannot initialize cov_factor before initializing scale.")
+        self._init_cov = scale
+
+    def get_posterior(self, *args, **kwargs):
+        """
+        Returns a posterior distribution over the latent unconstrained value.
+        """
+        loc = pyro.param("{}_loc".format(self.prefix), self._init_loc)
+        if self._init_scale == 0.:
+            return dist.Delta(loc, event_dim=1)  # Stage 1.
+        assert self._init_scale > 0.
+
+        @torch.no_grad()
+        def init_scale_fn():
+            # This heuristic aims to add proportional noise to values near 1 and
+            # to huge values, but will add overly large noise to tiny values.
+            data_scale = 1. + loc * loc
+            return data_scale * self._init_scale
+
+        scale = pyro.param("{}_scale".format(self.prefix), init_scale_fn,
+                           constraint=constraints.positive)
+        if self._init_cov == 0.:
+            return dist.Normal(loc, scale).to_event(1)  # Stage 2.
+        assert self._init_cov > 0.
+
+        @torch.no_grad()
+        def init_cov_factor_fn():
```

Comment:
not sure if i follow the logic of this heuristic

Reply:
Well there was an error :smile: I've added some comments.